src/validation.cpp

```diff
@@ -5191,6 +5191,282 @@ std::optional<std::reference_wrapper<const AssumeutxoData>> ExpectedAssumeutxo(
+{
+    uint256 base_blockhash = metadata.m_base_blockhash;
+
+    if (this->SnapshotBlockhash()) {
+        LogPrintf("[snapshot] can't activate a snapshot-based chainstate more than once\n");
+        return false;
+    }
+
+    int64_t current_coinsdb_cache_size{0};
+    int64_t current_coinstip_cache_size{0};
+
+    // Cache percentages to allocate to each chainstate.
+    //
+    // These particular percentages don't matter so much since they will only be
+    // relevant during snapshot activation; caches are rebalanced at the conclusion of
+    // this function. We want to give (essentially) all available cache capacity to the
+    // snapshot to aid the bulk load later in this function.
+    static constexpr double IBD_CACHE_PERC = 0.01;
+    static constexpr double SNAPSHOT_CACHE_PERC = 0.99;
+
+    {
+        LOCK(::cs_main);
+        // Resize the coins caches to ensure we're not exceeding memory limits.
+        //
+        // Allocate the majority of the cache to the incoming snapshot chainstate, since
+        // (optimistically) getting to its tip will be the top priority. We'll need to call
+        // `MaybeRebalanceCaches()` once we're done with this function to ensure
+        // the right allocation (including the possibility that no snapshot was activated
+        // and that we should restore the active chainstate caches to their original size).
+        //
+        current_coinsdb_cache_size = this->ActiveChainstate().m_coinsdb_cache_size_bytes;
+        current_coinstip_cache_size = this->ActiveChainstate().m_coinstip_cache_size_bytes;
+
+        // Temporarily resize the active coins cache to make room for the newly-created
+        // snapshot chain.
+        this->ActiveChainstate().ResizeCoinsCaches(
+            static_cast<size_t>(current_coinstip_cache_size * IBD_CACHE_PERC),
+            static_cast<size_t>(current_coinsdb_cache_size * IBD_CACHE_PERC));
+    }
+
+    auto snapshot_chainstate = WITH_LOCK(::cs_main, return MakeUnique<CChainState>(
+            this->ActiveChainstate().m_mempool, m_blockman, base_blockhash));
+
+    {
+        LOCK(::cs_main);
+        snapshot_chainstate->InitCoinsDB(
+            static_cast<size_t>(current_coinsdb_cache_size * SNAPSHOT_CACHE_PERC),
+            in_memory, false, "chainstate");
+        snapshot_chainstate->InitCoinsCache(
+            static_cast<size_t>(current_coinstip_cache_size * SNAPSHOT_CACHE_PERC));
+    }
+
+    const bool snapshot_ok = this->PopulateAndValidateSnapshot(
+        *snapshot_chainstate, coins_file, metadata);
+
+    if (!snapshot_ok) {
+        WITH_LOCK(::cs_main, this->MaybeRebalanceCaches());
+        return false;
+    }
+
+    {
+        LOCK(::cs_main);
+        assert(!m_snapshot_chainstate);
+        m_snapshot_chainstate.swap(snapshot_chainstate);
+        const bool chaintip_loaded = m_snapshot_chainstate->LoadChainTip(::Params());
+        assert(chaintip_loaded);
+
+        m_active_chainstate = m_snapshot_chainstate.get();
```

Comment:
In commit "validation: add ChainstateManager::ActivateSnapshot" (c5154cad6c195759409d6eb5134fad635b9aaade)

Credit to @jnewbery for pointing this out in https://github.com/bitcoin/bitcoin/pull/20749#discussion_r560979259, but the `ChainStateManager::ActiveChainstate` accessor isn't currently [locking cs_main](https://github.com/bitcoin/bitcoin/blob/e77eea5150bd46d2972ec5086b86948df6e73a83/src/validation.cpp#L5470-L5474) or [requiring cs_main to be locked](https://github.com/bitcoin/bitcoin/blob/e77eea5150bd46d2972ec5086b86948df6e73a83/src/validation.h#L887) to be locked. I think it needs to do one of these things now that the `m_active_chainstate` pointer can change here. It might also be appropriate to add GUARDED_BY annotation to `m_active_chainstate`.

Reply:
I'm most likely missing something, but I think it'd be sufficient just to lock cs_main in the ActiveChainstate accessor and not burden callers with locking `cs_main`. That way, callers of ActiveChainstate have the choice of either consistently using the chainstate reference they get, or locking cs_main themselves if they want to make sure that the chainstate doesn't change from under them.

Incidentally, it seems to me that locking `cs_main` in ActiveChainstate may prevent a race condition between the assertion and the return in `ActiveChainstate`

Reply:
I think eventually we'd like to move away from recursive locking, but that's a long way off for `cs_main`, so there's probably no harm in adding it here. Ultimately both are fine - as long as `cs_main` is held whenever `m_active_chainstate` is read or written to, then we're safe. One thing that you'd need to check for is that none of the callers of the various `Active*()` functions are holding a lower lock (e.g. `mempool.cs` or `g_cs_orphans`) and not `cs_main`. If so, then adding a `LOCK(cs_main)` to `ActiveChainstate()` would be a lock inversion.

Reply:
There appear to be quite a few other locations where `m_active_chainstate` is read/written without holding `cs_main`. One example: https://github.com/bitcoin/bitcoin/blob/52d84a45e2fa3def71853cd31d5582ae31ea85d3/src/validation.cpp#L5145-L5151

This violates the conditions in the comment: "This means it is safe to acquire the contents of this pointer with ::cs_main held, release the lock, and then use the reference without concern of it being deconstructed."

The earlier part of that comment also seems wrong after this PR: "Once this pointer is set to a corresponding chainstate, it will not be reset until init.cpp:Shutdown()." After this PR, the pointer is reset to a different value in `ActivateSnapshot()`.

I haven't been following the AssumeUTXO project, so perhaps this really is safe once snapshot activation is exposed, but from a quick code read it looks like there's the potential for a data race on `m_active_chainstate`.

Reply:
> This violates the conditions in the comment: "This means it is safe to acquire the contents of this pointer with ::cs_main held, release the lock, and then use the reference without concern of it being deconstructed."

How so? You can still acquire the pointer and use it without fear of segfault, since the underlying data will always correspond to either `m_idb_chainstate` or `m_snapshot_chainstate`. It may not be semantically what you want (though it may be too), but it will not segfault.

> The earlier part of that comment also seems wrong after this PR

Yes, this does seem out of date.

Reply:
> the underlying data will always correspond to either m_idb_chainstate or m_snapshot_chainstate

Not necessarily - it's not safe to read the pointer value when it's being written to by another thread since pointers aren't guaranteed to be atomic https://stackoverflow.com/questions/8919818/is-pointer-assignment-atomic-in-c